### PR TITLE
Allow requests to override automatic visibility

### PR DESCRIPTION
**Pull request by tjprescott**
Fix https://github.com/Azure/typespec-azure/issues/3214

### CODE REVIEW

File: packages/http/src/decorators.ts
Comment (tjprescott) [inline packages/http/src/decorators.ts:685]: Couldn't reuse the name `getRequestVisibility` because it conflicts with the one with verb as a param.

File: packages/http/src/decorators.ts
Comment (tjprescott) [inline packages/http/src/decorators.ts:690]: I wonder if it makes sense to try and get the verb here or if that would not be advisable due to decoration ordering?

File: packages/http/src/metadata.ts
Comment (tjprescott) [inline packages/http/src/metadata.ts:154]: This becomes the quintessential method you would want to call. I'm not sure if there is a good way to dissuade people from calling `getRequestVisibility` and `getOperationRequestVisibility` directly.

File: packages/samples/test/output/visibility/openapi.yaml
Comment (tjprescott) [inline packages/samples/test/output/visibility/openapi.yaml:216]: Hmm.... I don't think this is intended.

File: packages/samples/test/output/visibility/openapi.yaml
Comment (tjprescott) [inline packages/samples/test/output/visibility/openapi.yaml:242]: But this does seem correct. If these are updatable properties, they should be optional, not required, yes?

**Comment by bterlson** [comment]
This feels like something we could discuss briefly in a design meeting. To that end I have two questions: should we consider making a general purpose `inputVisibility` decorator in the compiler, which sets the visibility we apply for input types of an operation? And should we also have `outputVisibility` to specify same for the output types?

I feel like this could layer nicely with custom lifecycle methods, such that e.g. my JSON RPC `@query` lifecycle decorator can dispatch to `inputVisibility/outputVisibility` thus allowing other to understand what visibility rules to apply to the operation.

File: packages/samples/test/output/visibility/openapi.yaml
Comment (nguerrera) [reply packages/samples/test/output/visibility/openapi.yaml:242]: This is the same as above, array elements should keep their required fields.

File: packages/samples/test/output/visibility/openapi.yaml
Comment (nguerrera) [reply packages/samples/test/output/visibility/openapi.yaml:216]: I know what happened here. The Item flag should overrule the Patch flag because when you patch an array you have to replace the whole array and so the items inside the array have to be complete with a full set of required properties. Before the item flag did this because we checked visibility is Update and only update with no flags to make optional. 

(I haven't reviewed the whole PR yet.)

File: packages/http/src/metadata.ts
Comment (nguerrera) [reply packages/http/src/metadata.ts:154]: Again, haven't looked at everything, but we could not export getOperationRequestVisibility or mark it `/** @internal */` since it hasn't existed before. `getRequestVisibility(verb)` could be marked `/** @deprecated Use resolveRequestVisibility */`. If we need to use it under the covers, maybe move the body to an internal function with a new name like getRequestVisbilityForVerb() and have the deprecated function with the old name be a stub that calls that.

Comment (tjprescott) [reply packages/http/src/metadata.ts:154]: Well, I exported it because when we have a decorator we usually export the "getDecorator" variant... I just couldn't use the same name because we already export something with the name.

But realistically, in most cases I think you would want the logic of `resolveRequestVisibility` which essentially checks the `@requestVisibility` decorator first and falls back to the verb if needed.

File: packages/samples/test/output/visibility/openapi.yaml
Comment (tjprescott) [reply packages/samples/test/output/visibility/openapi.yaml:216]: Perfect. I'll write a test case for this

Comment (tjprescott) [reply packages/samples/test/output/visibility/openapi.yaml]: Okay, fixed.

File: packages/samples/test/output/visibility/openapi.yaml
Comment (tjprescott) [reply packages/samples/test/output/visibility/openapi.yaml]: Fixed.

File: packages/http/src/metadata.ts
Comment (nguerrera) [reply packages/http/src/metadata.ts:155]: Ah, I see. I think it would be OK to not export the 1:1 decorator getter since there's a good chance of using it instead of the thing that takes everything else into account. We could always export it later if it turns out that it's required outside the library for something, which I doubt will happen. I also don't feel too strongly on any of this, and it would be fine to me, for example, to just add to the doc comments of the other functions to say `you probably want {@link resolveRequestVisibility}`

File: packages/http/src/metadata.ts
Comment (nguerrera) [inline packages/http/src/metadata.ts:497]: I would update this to something like "Properties are made optional on patch requests if the visibility includes update, but not for array elements with item flag since you need to provide whole array elements with required properties even in a patch."

**Review by nguerrera — COMMENTED** [review]
Looks good, this is the approach we talked about and I only have minor implementation comments. 

I'm holding off on officially approving for now in case we do decide to have a design chat first as Brian suggested.

**Comment by nguerrera** [comment]
Worth talking about #2150 at the same time as this.

**Review by markcowl — CHANGES_REQUESTED** [review]
Technical: Before merging, I think we should have at least one sample that follows this patter in the Azure repo, we should also have a PR against the azure repo before we consider merging this here.

Philosophical: I agree that we should have a design meeting about this - while it is true that the 'create' visibility property does appear sometimes in the PATCH, it also doesn't appear when the PATCH is an update.  We need to ensure that this is the way we want to model CreateOnly properties, or whether we want to suggest having a separate API for creating secrets (enabling separate authorization for secrets management versus object creation).

I am not entirely sure that 'createOnly' is a pattern we want to support in standard Azure APIs, or what the right way to write separate create and update request schemas for PATCH might be given the limitations of OpenAPI

**Comment by tjprescott** [comment]
Closing in favor of https://github.com/microsoft/typespec/issues/2186

**Pull request closed without merging**
